### PR TITLE
[Core][ref-counter] expose nested_id's owner as part of return

### LINKED ISSUE #17553: [Core] ref count safety issue when task argument is nested in the return value.
sometimes OBOD gets lookups for objects that the ref counter thinks is out of scope, and responds with error. This causes ObjectNotFound to get raised to the user, crashing the problem:
```
[2021-08-02 18:38:01,396 D 18578 19138] reference_count.cc:500: Calling on_delete for object e2362ccb840337beffffffffffffffffffffffff0100000001000000
[2021-08-02 18:38:01,396 D 18578 19138] core_worker.cc:2626: Object e2362ccb840337beffffffffffffffffffffffff0100000001000000 is deleted. Unpinning the object.
[2021-08-02 18:38:01,396 D 18578 19138] reference_count.cc:483: Deleting Reference to object e2362ccb840337beffffffffffffffffffffffff0100000001000000
[2021-08-02 18:38:01,396 D 18578 19138] reference_count.cc:486: Releasing lineage for object e2362ccb840337beffffffffffffffffffffffff0100000001000000
[2021-08-02 18:38:01,396 D 18578 19138] task_manager.cc:411: Plasma object e2362ccb840337beffffffffffffffffffffffff0100000001000000 out of scope
[2021-08-02 18:38:01,396 D 18578 19138] task_manager.cc:413: Task e2362ccb840337beffffffffffffffffffffffff01000000 has e2362ccb840337beffffffffffffffffffffffff0100000001000000 in scope
[2021-08-02 18:38:01,396 D 18578 19138] task_manager.cc:416: Task e2362ccb840337beffffffffffffffffffffffff01000000 now has 0 plasma returns in scope
[2021-08-02 18:38:01,718 D 18578 19138] reference_count.cc:979: Tried to remove an object location for an object e2362ccb840337beffffffffffffffffffffffff0100000001000000 that doesn't exist in the reference table. It can happen if the object is already evicted.
[2021-08-02 18:38:02,406 D 18578 19138] reference_count.cc:1168: Tried to register a location subscriber for an object e2362ccb840337beffffffffffffffffffffffff0100000001000000 that doesn't exist in the reference table. The object has probably already been freed.
[2021-08-02 18:38:02,407 D 18578 19138] reference_count.cc:1168: Tried to register a location subscriber for an object e2362ccb840337beffffffffffffffffffffffff0100000001000000 that doesn't exist in the reference table. The object has probably already been freed.
```

repro cluster: https://beta.anyscale.com/o/anyscale-internal/projects/prj_Rztp4a5rBZCDvG6vtLZDgpX8/?tab=session-list

```
import time
import numpy as np
import torch
import ray
import argparse
paths = [
    f's3://shuffling-data-loader-benchmarks/data/input_data_{i}.parquet.snappy' for i in range(0, 25)
]
def create_parser():
    parser = argparse.ArgumentParser(description="Eric Example")
    parser.add_argument("--address")
    parser.add_argument(
        "--batch-size",
        type=int,
        default=250000,
        metavar="N",
        help="input batch size for training (default: 64)")
    parser.add_argument("--num-workers", type=int, default=4)
    return parser
def create_torch_iterator(split, batch_size, rank=None):
    print(f"Creating Torch shuffling dataset for worker {rank} with "
          f"{batch_size} batch size.")
    numpy_to_torch_dtype = {
        np.bool: torch.bool,
        np.uint8: torch.uint8,
        np.int8: torch.int8,
        np.int16: torch.int16,
        np.int32: torch.int32,
        np.int64: torch.int64,
        np.float16: torch.float16,
        np.float32: torch.float32,
        np.float64: torch.float64,
        np.complex64: torch.complex64,
        np.complex128: torch.complex128
    }
    DATA_SPEC = {
        "embeddings_name0": (0, 2385, np.int64),
        "embeddings_name1": (0, 201, np.int64),
        "embeddings_name2": (0, 201, np.int64),
        "embeddings_name3": (0, 6, np.int64),
        "embeddings_name4": (0, 19, np.int64),
        "embeddings_name5": (0, 1441, np.int64),
        "embeddings_name6": (0, 201, np.int64),
        "embeddings_name7": (0, 22, np.int64),
        "embeddings_name8": (0, 156, np.int64),
        "embeddings_name9": (0, 1216, np.int64),
        "embeddings_name10": (0, 9216, np.int64),
        "embeddings_name11": (0, 88999, np.int64),
        "embeddings_name12": (0, 941792, np.int64),
        "embeddings_name13": (0, 9405, np.int64),
        "embeddings_name14": (0, 83332, np.int64),
        "embeddings_name15": (0, 828767, np.int64),
        "embeddings_name16": (0, 945195, np.int64),
        "one_hot0": (0, 3, np.int64),
        "one_hot1": (0, 50, np.int64),
        "labels": (0, 1, np.float64),
    }
    feature_columns = list(DATA_SPEC.keys())
    feature_types = [
        numpy_to_torch_dtype[dtype] for _, _, dtype in DATA_SPEC.values()
    ]
    label_column = feature_columns.pop()
    label_type = feature_types.pop()
    torch_iterator = split.to_torch(
         label_column=label_column,
         feature_columns=feature_columns,
         label_column_dtype=label_type,
         feature_column_dtypes=feature_types,
         batch_size=batch_size,
         # prefetch_blocks: int = 0,
         # drop_last: bool = False
    )
    return torch_iterator
def create_dataset(filenames):
    pipeline = ray.data.read_parquet(list(filenames))\
        .repeat().random_shuffle()
    return pipeline
if __name__ == '__main__':
    parser = create_parser()
    args = parser.parse_args()
    print("Connecting to Ray cluster...")
    ray.init(address=args.address)
    pipeline = create_dataset(paths)
    splits = pipeline.split(args.num_workers)
    @ray.remote
    def consume(split, rank=None, batch_size=None):
        torch_iterator = create_torch_iterator(split, batch_size=batch_size, rank=rank)
        for i, (x, y) in enumerate(torch_iterator):
            time.sleep(1)
            if i % 10 == 0:
                print(i)
        return
    tasks = [
        consume.remote(split, rank=idx, batch_size=args.batch_size)
        for idx, split in enumerate(splits)
    ]
    ray.get(tasks)
```

### PR DESCRIPTION
**Pull request by scv119**
hmm not sure if this PR is relevant to the bug given we found the fix in #17802

### CODE REVIEW

**Review by stephanie-wang — COMMENTED** [review]
This looks like the right fix to me!

The only thing is that I think we need to extend the `nested_inlined_ids` field to include those objects' owner addresses instead, since this info needs to be added into the ref table on the borrower.

File: src/ray/core_worker/reference_count.cc [hunk 1]
```diff
@@ -818,6 +818,28 @@ void ReferenceCounter::AddNestedObjectIds(const ObjectID &object_id,
   AddNestedObjectIdsInternal(object_id, inner_ids, owner_address);
 }
 
+void ReferenceCounter::UpdateOwnedObjectNestedObjectIds(
```
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.cc:821]: There is an `AddNestedObjectIds` method that we should probably modify/use instead.

Also, we need to pass the inner IDs' owner addresses here.

**Comment by stephanie-wang** [comment]
Interestingly, the doc for `AddNestedObjectIds` actually seems to describe the exact scenario we're running into, but it doesn't actually seem to get called.

```
  /// Mark that this ObjectID contains another ObjectID(s). This should be
  /// called in two cases:
  /// 1. We are storing the value of an object and the value contains
  /// serialized copies of other ObjectIDs. If the outer object is owned by a
  /// remote process, then they are now a borrower of the nested IDs.
  /// 2. We submitted a task that returned an ObjectID(s) in its return values
  /// and we are processing the worker's reply. In this case, we own the task's
  /// return objects and are borrowing the nested IDs.
```

Looks like 1 is happening in core_worker.cc but not 2.

**Comment by rkooo567** [comment]
It is not so common case, so it is probably slipped in implementation haha

**Pull request closed without merging**